### PR TITLE
Add test for copy-to-clipboard functionality in `passcode` page

### PR DESCRIPTION
**Pull request by hsinn0**
[#183422195]

### CODE REVIEW

**Review by strehle — APPROVED** [review]
Thanks

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java [hunk 1]
```diff
@@ -269,6 +269,13 @@ public void testPasscodeRedirect() {
         attemptLogin(testAccounts.getUserName(), testAccounts.getPassword());
 
         assertThat(webDriver.findElement(By.cssSelector("h1")).getText(), Matchers.containsString("Temporary Authentication Code"));
+
+        // Verify that the CopyToClipboard function can be executed
+        String passcode = webDriver.findElement(By.id("passcode")).getText();
+        ((JavascriptExecutor)webDriver).executeScript("CopyToClipboard",
+                passcode);
+        // Verify that the copybutton can be clicked
+        webDriver.findElement(By.id("copybutton")).click();
```
Comment (bruce-ricard) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:278]: Not knowing what the semantics of the `click()` method is, from its name it looks to me that it just clicks on the element. But you can always click on anything, even if it's not per say clickable. It seems to me like what we should be testing is that the clicking actually has some impact. Maybe could we test that passcode is indeed in the clipboard?

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java [hunk 2]
```diff
@@ -269,6 +269,13 @@ public void testPasscodeRedirect() {
         attemptLogin(testAccounts.getUserName(), testAccounts.getPassword());
 
         assertThat(webDriver.findElement(By.cssSelector("h1")).getText(), Matchers.containsString("Temporary Authentication Code"));
+
+        // Verify that the CopyToClipboard function can be executed
```
Comment (bruce-ricard) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:273]: It seems to me like this test is mixing 2 different layers of abstraction, by testing the javascript method, and then testing the UI.

This is an integration test, it seems to me like we could just test that the functionality is working, when we click on the button.

**Review by bruce-ricard — CHANGES_REQUESTED** [review]

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 1 above)
Comment (hsinn0) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:278]: The button click will copy the passcode to the OS clipboard.  Not something I want to check in Java code.  I also checked if I can check for some kind of event, but was not feasible without changing the involved javascript function.
This case, at least, will test the existence of the 'copybutton'.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 2 above)
Comment (hsinn0) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:273]: By testing both the button's existence and it is clickable, and also the Java function is loaded and can be executed, we at least test both components involved in the action.  I do understand that we miss the part verifying that the function is being executed when you click the button.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java (hunk 1 above)
Comment (bruce-ricard) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/LoginIT.java:278]: It looks like it's fairly easy to get the contents of the clipboard: https://stackoverflow.com/a/7106086/1296753

How would you feel about testing the contents of the clipboard just after the click, to validate that it contains the passcode?

**Pull request merged**